go/store/prolly/tree/merge.go

```diff
@@ -88,172 +81,295 @@ func ThreeWayMerge[K ~[]byte, O Ordering[K], S message.Serializer](
+func SendPatches[K ~[]byte, O Ordering[K]](
 	ctx context.Context,
-	l, r Differ[K, O],
+	l, r PatchGenerator[K, O],
 	buf PatchBuffer,
 	cb CollisionFn,
-) (stats MergeStats, err error) {
+) (err error) {
 	var (
-		left, right Diff
-		lok, rok    = true, true
+		left, right          Patch
+		lDiffType, rDiffType DiffType
+		lok, rok             = true, true
 	)
 
-	left, err = l.Next(ctx)
+	left, lDiffType, err = l.Next(ctx)
 	if err == io.EOF {
 		err, lok = nil, false
 	}
 	if err != nil {
-		return MergeStats{}, err
+		return err
 	}
 
-	right, err = r.Next(ctx)
+	right, rDiffType, err = r.Next(ctx)
 	if err == io.EOF {
 		err, rok = nil, false
 	}
 	if err != nil {
-		return MergeStats{}, err
+		return err
 	}
 
 	for lok && rok {
-		cmp := l.order.Compare(ctx, K(left.Key), K(right.Key))
+		order := l.order
+		// If they're ranges, compare the start points, see if they overlap.
+		leftLevel, _ := l.getLevel()
+		rightLevel, _ := r.getLevel()
+		if leftLevel > 0 && rightLevel > 0 {
+			if nilCompare(ctx, order, K(left.EndKey), K(right.KeyBelowStart)) <= 0 {
+				// Left change is entirely before right change.
+				// This change is already on the left map, so we ignore it.
+				left, lDiffType, err = l.Next(ctx)
+				if err == io.EOF {
+					err, lok = nil, false
+				}
+				if err != nil {
+					return err
+				}
+			} else if nilCompare(ctx, order, K(right.EndKey), K(left.KeyBelowStart)) <= 0 {
+				// Right change is entirely before right change.
+				err = buf.SendPatch(ctx, right)
+				if err != nil {
+					return err
+				}
+
+				right, rDiffType, err = r.Next(ctx)
+				if err == io.EOF {
+					err, rok = nil, false
+				}
+				if err != nil {
+					return err
+				}
+			} else if bytes.Equal(left.To, right.To) {
+				// A concurrent change.
+				// This change is already on the left map, so we ignore it.
+				left, lDiffType, err = l.Next(ctx)
+				if err == io.EOF {
+					err, lok = nil, false
+				}
+				if err != nil {
+					return err
+				}
+
+				right, rDiffType, err = r.Next(ctx)
+				if err == io.EOF {
+					err, rok = nil, false
+				}
+				if err != nil {
+					return err
+				}
+			} else {
+				// In all other cases there's a conflict and we have to split whichever one comes first.
+				// If both have the same start key, split both.
+				cmp := nilCompare(ctx, order, K(left.KeyBelowStart), K(right.KeyBelowStart))
+				if cmp <= 0 {
+					left, lDiffType, err = l.split(ctx)
+					if err != nil {
+						return err
+					}
+				}
+				if cmp >= 0 {
+					right, rDiffType, err = r.split(ctx)
+					if err != nil {
+						return err
+					}
+				}
+			}
+			continue
+		}
+
+		// If one branch returns a range patch and the other returns a point patch, we need to see if they overlap and possibly split the range diff.
+		if rightLevel > 0 {
+			if nilCompare(ctx, l.order, K(left.EndKey), K(right.KeyBelowStart)) <= 0 {
```

Comment:
When do we use `order` and when do we use `l.order`? Should be `order` everywhere, right?